homeassistant/components/google_calendar.py

```diff
@@ -0,0 +1,407 @@
+    import os
+
+    config_file = hass.config.path(CONFIG_FILE)
+    if not os.path.isfile(config_file):
+        # we don't have a setup yet
+        request_api_setup(hass, config)
+        return False
+
+    # check that oauth is setup
+    token_file = hass.config.path(TOKEN_FILE)
+    if not os.path.isfile(token_file):
+        # we don't have a setup yet
+        request_oauth_setup(hass, config)
+        return False
+
+    return True
+
+
+def setup(hass, config):
+    """Setup the platform."""
+    config = dict(config)  # copy it so we can modify
+
+    config = config.get(DOMAIN, {})
+    if isinstance(config, list) and len(config) > 0:
+        config = config[0]
+
+    def _scan_for_calendars(service):
+        """Scan for new calendars."""
+        from homeassistant.util import convert
+        track_new = convert(config.get(CONF_TRACK_NEW),
+                            bool, DEFAULT_CONF_TRACK_NEW)
+        service = get_calendar_service(hass)
+        calendars = service.calendarList().list().execute()['items']
+        for calendar in calendars:
+            found_calendar(hass, calendar, track_new)
+
+    hass.services.register(
+        DOMAIN, SERVICE_SCAN_CALENDARS, _scan_for_calendars,
+        None, schema=None)
+
+    if not do_setup_check(hass, config):
+        return False
+
+    do_setup(hass, config)
+    return True
+
+
+def do_setup(hass, config):
+    """Run the setup after we have everything configured."""
+    # pylint disable=global-statement
+    from homeassistant.loader import get_component
+
+    if DOMAIN in _CONFIGURING:
+        get_component('configurator').request_done(_CONFIGURING.pop(DOMAIN))
+
+    # Ensure component is loaded
+    bootstrap.setup_component(hass, 'binary_sensor', config)
+
+    calendars = load_config(hass.config.path(YAML_DEVICES))
+    for calendar_hash, calendar in calendars.items():
+        _CALENDARS.update({calendar_hash: calendar})
+        platform_discovered(hass, calendar_hash)
+
+    hass.services.call(DOMAIN, SERVICE_SCAN_CALENDARS, None)
+    return True
+
+
+def found_calendar(hass, calendar, track_new):
+    """Check if we know about a calendar and generate PLATFORM_DISCOVER."""
+    # since we could just use the global and not assign it
+    import hashlib
+
+    calendar_hash = hashlib.sha224(
+        calendar['id'].encode('utf-8')).hexdigest()
+
+    if _CALENDARS.get(calendar_hash, None) is not None:
+        return
+
+    # seriously, FU pylint
+    _CALENDARS.update({calendar_hash: {
+        CONF_CAL_ID: calendar['id'],
+        CONF_NAME: calendar['summary'],
+        CONF_ENTITIES: [{
+            CONF_TRACK: track_new,
+            CONF_SEARCH: None,
+            CONF_NAME: calendar['summary']
+        }]
+    }})
+
+    update_config(
+        hass.config.path(YAML_DEVICES),
+        calendar_hash,
+        _CALENDARS[calendar_hash]
+    )
+
+    platform_discovered(hass, calendar_hash)
+
+
+def platform_discovered(hass, calendar_hash):
```

Comment:
Instead of adding this method, please use `discover()` in the `discovery` component.